pkg/controller/persistentvolume/index.go

```diff
@@ -75,27 +74,38 @@ func (pvIndex *persistentVolumeOrderedIndex) findByClaim(claim *api.PersistentVo
 	// potential matches (the GCEPD example above).
 	allPossibleModes := pvIndex.allPossibleMatchingAccessModes(claim.Spec.AccessModes)
 
+	var smallestVolume *api.PersistentVolume
+	var smallestVolumeSize int64
+	requestedQty := claim.Spec.Resources.Requests[api.ResourceName(api.ResourceStorage)]
+	requestedSize := requestedQty.Value()
+
 	for _, modes := range allPossibleModes {
 		volumes, err := pvIndex.listByAccessModes(modes)
 		if err != nil {
 			return nil, err
 		}
 
-		// volumes are sorted by size but some may be bound or earmarked for a specific claim.
-		// filter those volumes for easy binary search by size
-		// return the exact pre-binding match, if found
-		unboundVolumes := []*api.PersistentVolume{}
+		// Go through all available volumes and find the smallest matching one.
+		// Return the exact pre-binding match, if found.
 		for _, volume := range volumes {
-			// volume isn't currently bound or pre-bound.
-			if volume.Spec.ClaimRef == nil {
-				unboundVolumes = append(unboundVolumes, volume)
-				continue
-			}
-
 			if isVolumeBoundToClaim(volume, claim) {
 				// Exact match! No search required.
 				return volume, nil
 			}
+
+			if volume.Spec.ClaimRef != nil {
```

Comment:
I have absolutely no doubt that **this** will speed things upsignificantly.  I was in this code today and was wondering if perhaps where along the line an if statement like this was lost, because the code says it's getting unbound volumes, when in fact it makes no check for binding except for the case where a volume is pre-bound, so we're essentially scanning the **entire** set of volumes right now and then applying a sort.